### PR TITLE
Rework transactions

### PR DESCRIPTION
**Pull request by Tyrrrz**

### CODE REVIEW

**Comment by Tyrrrz** [comment]
@bruno-garcia RFC

File: src/Sentry/Extensibility/DisabledHub.cs
Comment (bruno-garcia) [inline src/Sentry/Extensibility/DisabledHub.cs:74]: Ideally we'd keep this `ISpan`. Technically any span can be set to the scope, you can create a child and set to the scope so next startSpan would be two levels down from the transaction

File: src/Sentry/Extensibility/HubAdapter.cs
Comment (bruno-garcia) [inline src/Sentry/Extensibility/HubAdapter.cs:77]: This should return `ISpan`

File: src/Sentry/ISpanTracer.cs
Comment (bruno-garcia) [inline src/Sentry/ISpanTracer.cs:31]: You should be able to create a child from a span. Why did this need to move to this interface?

**Comment by bruno-garcia** [comment]
Turned out quite different than Java and Objective-C, could you please elaborate the motivations?

File: src/Sentry/Extensibility/DisabledHub.cs
Comment (Tyrrrz) [reply src/Sentry/Extensibility/DisabledHub.cs:74]: `ISpan` doesn't have `StartChild()` though so we do need `ISpanTracer` here tho. Unless you have something else in mind?

**Comment by Tyrrrz** [comment]
@bruno-garcia it's actually the same as Java. The only difference is that I added interfaces for the Transaction/Span tracers. Other than that it's identical.

File: src/Sentry/ISpanTracer.cs
Comment (Tyrrrz) [reply src/Sentry/ISpanTracer.cs:31]: Span effectively cannot be modified, as it represents the final state included in a transaction. It's the same as Java's `SentrySpan` vs `Span`.

File: src/Sentry/Extensibility/DisabledHub.cs
Comment (bruno-garcia) [reply src/Sentry/Extensibility/DisabledHub.cs:74]: Can't we add StartChild on `ISpan`? A Span can have children so I don't see a reason not to:
Java has it too: https://github.com/getsentry/sentry-java/blob/7c3f88ad24c95af317c73a042951e1eabc4032be/sentry/src/main/java/io/sentry/Span.java#L67

If we really wanted to split the serialization format from what we have floating around at runtime it would be a reason not to add this. But `ISpanTracer` as a public type is a new concept that wasn't discussed with the wider team.

Comment (bruno-garcia) [reply src/Sentry/Extensibility/DisabledHub.cs:74]: @marandaneto ^

File: src/Sentry/ISpanTracer.cs
Comment (bruno-garcia) [reply src/Sentry/ISpanTracer.cs:31]: Isn't SpanContext cannot be modified?

**Comment by bruno-garcia** [comment]
This PR includes a lot of breaking changes so before we go ahead and merge this (and essentially sign up for releasing 4.0.0 already) lets make sure we get all the details about this right.

**Comment by Tyrrrz** [comment]
@bruno-garcia how should we go around it? set up a meeting or wait for the usual sync?

**Comment by bruno-garcia** [comment]
Ideally we keep this rolling if just async until we can make a decision. 
Let me ping folks on Discord

File: src/Sentry/Extensibility/DisabledHub.cs
Comment (marandaneto) [reply src/Sentry/Extensibility/DisabledHub.cs:74]: yep, sounds like `ISpan` would do the job

File: src/Sentry/Extensibility/HubAdapter.cs
Comment (marandaneto) [reply src/Sentry/Extensibility/HubAdapter.cs:77]: on Java, `StartTransaction` still returns `ITransaction` which extends `ISpan` to not do breaking changes, `getSpan` returns `ISpan`, and `startChild` aways returns `ISpan`

that's what makes the most sense for me, in the feature `StartTransaction` could also return `ISpan`

right now this introduces a breaking change by changing the return type, we've agreed to not do that, right? unless `ITransactionTracer` implements  `ITransaction`, didn't check yet

File: src/Sentry/ISpanTracer.cs
Comment (marandaneto) [reply src/Sentry/ISpanTracer.cs:31]: on Java, `Span` takes a `SentryTracer` as ctor param and when `startChild` is called, it calls `SentryTracer#startChild`

File: src/Sentry/ISpanTracer.cs
Comment (marandaneto) [inline src/Sentry/ISpanTracer.cs:8]: ISpanTracer and ITransactionTracer add a suffix to the API, what was the inspiration for it?

File: src/Sentry/Span.cs
Comment (marandaneto) [inline src/Sentry/Span.cs:184]: I'm not a big fun of `!`, maybe check if its null and set a new Dic if null? just for the sake of nul safety

File: src/Sentry/TransactionTracer.cs
Comment (marandaneto) [inline src/Sentry/TransactionTracer.cs:31]: I'm not entirely sure if it's actually helpful to have this in the docs, we already have an issue for that, right?
giving examples of what to NOT do is..a red flag.

File: src/Sentry/TransactionTracer.cs
Comment (marandaneto) [inline src/Sentry/TransactionTracer.cs:245]: should we make this Atomic? it is on Java

**Comment by marandaneto** [comment]
so my 2 cents are pretty much what @Tyrrrz said, looks the same as Java except for the other 2 interfaces, and the reasoning is how the code is structured, immutability or not, etc.
so @bruno-garcia your call to either keep as it is or refactoring to get rid of them and keep as Java with ISpan ITransaction like, as I don't know the code base, I can't really give more than that.

File: src/Sentry/TransactionTracer.cs
Comment (Tyrrrz) [reply src/Sentry/TransactionTracer.cs:31]: yeah, this was added before the issue existed.
i will remove it.

File: src/Sentry/ISpanTracer.cs
Comment (Tyrrrz) [reply src/Sentry/ISpanTracer.cs:8]: I think the main difference in this version is that

ISpan - immutable; Protocol.Span implements ISpan
ISpanTracer - ISpan + mutations (including StartChild); SpanTracer implements ISpanTracer; potentially NoOpSpanTracer would also implement it.

and then essentially the same with ITransaction/ITransactionTracer.

File: src/Sentry/Span.cs
Comment (Tyrrrz) [reply src/Sentry/Span.cs:184]: I agree.
we can't fully guarantee that the map is not nullable in JSON, but we're responsible for both serialization and deserialization parts here, so we can reasonably guarantee it.
in general though, it would be better to add a proper check, but it's on our other list of improvements: #631

File: src/Sentry/TransactionTracer.cs
Comment (Tyrrrz) [reply src/Sentry/TransactionTracer.cs:245]: @marandaneto 
I noticed that the code inside the `clearTransaction()` method is locked, but the `if (transaction == this)` part is not:

https://github.com/getsentry/sentry-java/blob/6c371b38c494ba08d0f4fbe4823cde9f378f1b5f/sentry/src/main/java/io/sentry/SentryTracer.java#L106-L111

In C# we can do `Interlocked.CompareExchange(...)` to make an atomic compare + replace operation, but that requires a direct field access...

**Comment by Tyrrrz** [comment]
@marandaneto @bruno-garcia I adjusted the naming a bit so that it's less of a breaking change now.

File: src/Sentry/TransactionTracer.cs
Comment (marandaneto) [reply src/Sentry/TransactionTracer.cs]: @Tyrrrz `scope.withTransaction` is fully locked, have a look at the `withTransaction` impl

File: CHANGELOG.md
Comment (bruno-garcia) [inline CHANGELOG.md:7]: Please add a more descriptive entry such as `CaptureTransaction takes concrete Transaction class` or whatever else was done here worth noting to the users?

File: src/Sentry/Span.cs
Comment (bruno-garcia) [reply src/Sentry/Span.cs:184]: We had a few experiences with deserializing something that we expect to be in a certain format only to find out it's not.

I agree with @marandaneto here and even though not the nicest thing, ideally we'd:

`tags ?? new Dictionary<>()` to avoid a null reference exception if we're accessing `tags` from  a deserialized instance.

File: src/Sentry/Transaction.cs
Comment (bruno-garcia) [inline src/Sentry/Transaction.cs:184]: nit:

```suggestion
            : this(tracer.Name, tracer.Operation)
```

**Review by bruno-garcia — APPROVED** [review]
Great stuff! Thanks.
I left a couple of small notes before we merge

File: src/Sentry/Span.cs
Comment (Tyrrrz) [reply src/Sentry/Span.cs:184]: @bruno-garcia it's not the `tags` that is null here. The error is raised because the second generic argument in dictionary is nullable here.

Comment (bruno-garcia) [reply src/Sentry/Span.cs:184]: Oh, I see now. Not clear at all from just `!`. Lost chance to have a nicer syntax

`tags<>!` :P

**Review by bruno-garcia — APPROVED** [review]

**Comment by bruno-garcia** [comment]
@Tyrrrz LGTM but CI has a test timing out after 7 seconds. Flaky or bug?

File: src/Sentry/Span.cs
Comment (Tyrrrz) [reply src/Sentry/Span.cs:184]: Indeed, that would have been much better :P

**Comment by Tyrrrz** [comment]
@bruno-garcia yeah the same flaky tests are failing :/
I remember we briefly discussed that they could be caused by inconsistent thread locking in xUnit. I think we should look into it as soon as we can, it has become more common lately, even though the tests have not changed.

**Pull request merged**
